File: Platforms/MulticoreCPU/multicore_c_code/option.c
/*
 * option.c
 *
 *  Created on: 10 November 2012
 *      Author: gordon
 */
#include "option.h"

void option_derivative_init(double t,double c,double k,double p,option_opt_attr* o_a){
	o_a->strike_price = k;
	o_a->call=c;
	o_a->time_period=t;
	o_a->points = p;
}

void option_derivative_path_init(option_opt_var* o_v,option_opt_attr* o_a){
	o_v->value=0;
	o_v->delta_time=o_a->time_period/o_a->points;
}

void option_derivative_path(double price,double time,option_opt_var* o_v,option_opt_attr* o_a){
}

void option_derivative_payoff(double end_price,option_opt_var* o_v,option_opt_attr* o_a){
	if((o_a->call)){
		o_v->value = (end_price-o_a->strike_price);
	}
	else{
		o_v->value = (o_a->strike_price-end_price);
	}
}

File: Platforms/MaxelerFPGA/maxeler_code/src/mc_solver_maxeler/option.java
package mc_solver_maxeler;

import com.maxeler.maxcompiler.v1.kernelcompiler.Kernel;
import com.maxeler.maxcompiler.v1.kernelcompiler.KernelLib;
import com.maxeler.maxcompiler.v1.kernelcompiler.types.base.HWVar;

public class option extends KernelLib{
	//HWFloat doubleType = Kernel.hwFloat(11,53);
	String name = "option";

	protected option_parameters parameters;

	protected HWVar delta_time;
	protected HWVar carried_delta_time;
	protected HWVar new_delta_time;

	protected HWVar value;

	protected HWVar enable;
	protected HWVar carried_enable;
	protected HWVar new_enable;

	//protected HWVar temp_price;

	Kernel kernel;
	HWVar path;
	HWVar point;

	public option(MC_Solver_Maxeler_Base_Kernel kernel,HWVar pp,HWVar p,HWVar enable,option_parameters op) {
		super(kernel);
		this.kernel = kernel;
		this.point = pp;
		this.path = p;
		this.enable = enable;

		//this.temp_price = temp_price;

		this.parameters = op;
	}

	public void path_init(){
		carried_delta_time = ((MC_Solver_Maxeler_Base_Kernel)this.kernel).inputDoubleType.newInstance(((MC_Solver_Maxeler_Base_Kernel)this.kernel));

		this.delta_time = this.point.eq(0) ? ((this.parameters.time_period/(this.parameters.points))).cast(((MC_Solver_Maxeler_Base_Kernel)this.kernel).inputDoubleType) : this.carried_delta_time;
		}

	public void path(HWVar temp_price,HWVar time){
		this.new_delta_time = this.enable.eq(true)?this.parameters.time_period/(this.parameters.points):0.0;
		this.new_enable = this.enable;
		//this.value = (this.point.eq(this.kernel.path_points-1)) ? this.payoff(temp_price) : this.kernel.constant.var(this.kernel.doubleType,0.0);
	}

	public void connect_path(){
		this.carried_delta_time <== ((MC_Solver_Maxeler_Base_Kernel)this.kernel).stream.offset(this.new_delta_time,-((MC_Solver_Maxeler_Base_Kernel)this.kernel).instance_paths);
	}

	public HWVar payoff(HWVar end_price){
		return (this.parameters.call).eq(0) ? this.parameters.strike_price-end_price : end_price-this.parameters.strike_price;
	}

}
